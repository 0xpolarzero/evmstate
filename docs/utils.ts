import { isHex } from "tevm";
import clsx, { type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export const cn = (...inputs: ClassValue[]) => {
  return twMerge(clsx(inputs));
};

const maxBytesLength = 64;
export const stringify = (value: any, space: number = 2) => {
  return JSON.stringify(
    value,
    (_, v) => {
      if (typeof v === "bigint") return `__bigint__${v.toString()}`;
      if (isHex(v) && (v.length - 2) / 2 > maxBytesLength)
        return `__long_hex__${v.slice(0, maxBytesLength + 2)}...${v.slice(-maxBytesLength)}`;
      return v;
    },
    space,
  );
};
