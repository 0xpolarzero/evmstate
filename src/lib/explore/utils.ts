import { Hex, toHex } from "tevm";
import { ByteArray, ToHexParameters } from "viem";

export const max = <T extends bigint | number>(...nums: Array<T>): T =>
  // @ts-expect-error bigint/number
  nums.reduce((max, num) => (num > max ? num : max), -Infinity);

/**
 * Converts a value to a hex string and ensures it has an even number of characters (full bytes)
 *
 * @param value The value to convert to hex
 * @param options Optional configuration options passed to tevm's toHex
 * @returns A hex string with full bytes
 */
export const toHexFullBytes = (value: string | number | bigint | boolean | ByteArray, opts?: ToHexParameters): Hex => {
  // Use tevm's toHex function
  const hex = toHex(value, opts);

  // If the hex string has an odd number of characters after the 0x prefix,
  // pad it with a leading zero to make it even
  if ((hex.length - 2) % 2 !== 0) return ("0x0" + hex.slice(2)) as Hex;
  return hex;
};
