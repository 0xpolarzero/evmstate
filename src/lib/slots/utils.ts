import { Address, decodeAbiParameters, encodeAbiParameters, Hex, isHex, toHex } from "tevm";
import { abi } from "@shazow/whatsabi";
import { AbiType, AbiTypeToPrimitiveType } from "abitype";
import { padHex } from "viem";

import { MappingKey } from "@/lib/slots/types";

/* -------------------------------------------------------------------------- */
/*                                  MAPPINGS                                  */
/* -------------------------------------------------------------------------- */

/** Extract values from a transaction trace that might be used as mapping keys */
// TODO: refactor to navigate the stack trace and identify both potential keys and array indexes
/**
 * Typical access patterns:
 *
 * Mapping
 *
 * - KECCAK256 → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(abi.encode(key, uint256(mappingSlot)))
 * - -> key is the input to the hash so we might even be able to retrieve the slot and compare directly
 * - KECCAK256 → KECCAK256 → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(abi.encode(key2, keccak256(abi.encode(key1, uint256(mappingSlot)))))
 * - -> same here we get each key from each keccak input
 *
 * Array
 *
 * - KECCAK256 → ADD → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(uint256(arraySlot)) + index
 *
 * As reference:
 *
 * - Struct in array: KECCAK256 → ADD → ADD → SLOAD/SSTORE Slot calculation: (keccak256(uint256(arraySlot)) + index) +
 *   memberOffset
 * - Mapping to Struct with Mapping: KECCAK256 → ADD → KECCAK256 → SLOAD/SSTORE Slot calculation:
 *   keccak256(abi.encode(key2, (keccak256(abi.encode(key1, uint256(mappingSlot))) + memberOffset)))
 */
export const extractPotentialKeys = (
  trace: {
    uniqueStackValues?: Array<string>;
    relevantOps?: Array<{
      op: string;
      stack: Array<string>;
    }>;
  },
  addresses: Array<Address>,
  abiFunctions: Array<abi.ABIFunction>,
  txData?: Hex,
): MappingKey[] => {
  const keys: MappingKey[] = [];

  // Add touched addresses
  addresses.forEach((address) => {
    keys.push({
      hex: padHex(address, { size: 32 }),
      decoded: address,
      type: "address",
    });
  });

  // Extract parameters from transaction data
  if (txData && txData.length > 10) {
    const selector = txData.slice(0, 10);
    const inputs = abiFunctions.find((fn) => fn.selector === selector)?.inputs;

    if (inputs) {
      // Decode function inputs
      const params = decodeAbiParameters(inputs, `0x${txData.slice(10)}`);

      params.forEach((param, index) => {
        // If it's an array, add each element as a key
        if (Array.isArray(param)) {
          param.forEach((p) => {
            const type = inputs[index].type.replace("[]", "") as AbiType;

            if (type && type !== "string" && type !== "bytes") {
              keys.push({
                hex: padHex(encodeAbiParameters([{ type }], [p]), { size: 32 }),
                decoded: p as AbiTypeToPrimitiveType<typeof type>,
                type,
              });
            }
          });
        } else {
          // Otherwise just add the key straight up
          const type = inputs[index].type as AbiType;

          if (type && type !== "string" && type !== "bytes") {
            keys.push({
              hex: padHex(encodeAbiParameters([{ type }], [param]), { size: 32 }),
              decoded: param as AbiTypeToPrimitiveType<typeof type>,
              type,
            });
          }
        }
      });
    }
  }

  // Process stack values from the trace
  if (trace.uniqueStackValues?.length) {
    // Process unique stack values directly
    for (const stackValue of trace.uniqueStackValues) {
      keys.push({
        hex: isHex(stackValue) ? padHex(stackValue, { size: 32 }) : toHex(stackValue, { size: 32 }),
        type: undefined,
      });
    }
  }

  // Deduplicate keys
  const uniqueMap = new Map();
  // Add the new key only if it's not already in the map (and don't replace a key with a defined type)
  keys.forEach((k) => {
    if (!uniqueMap.has(k.hex) || k.type) uniqueMap.set(k.hex, k);
  });

  return Array.from(uniqueMap.values()).sort((a, b) => {
    // prefer address as it's more likely to be a key
    if (a.type === "address") return -1;
    if (b.type === "address") return 1;
    // prefer defined types
    if (a.type === undefined) return 1;
    if (b.type === undefined) return -1;

    return 0;
  });
};

/* -------------------------------------------------------------------------- */
/*                                    UTILS                                   */
/* -------------------------------------------------------------------------- */

/** A helper function to clean up trace objects by removing undefined or zero values */
export const cleanTrace = (obj: any) => {
  const { current, next, note, ...rest } = obj;
  let trace = { ...rest };

  // Only include note if it exists
  if (note) trace.note = note;

  // Same for current and next
  trace.current = { hex: current.hex };
  if (current.decoded !== undefined) trace.current = { hex: current.hex, decoded: current.decoded };

  if (next && rest.modified) {
    trace.next = { hex: next.hex };
    if (next.decoded !== undefined) trace.next = { hex: next.hex, decoded: next.decoded };
  }

  return trace;
};
