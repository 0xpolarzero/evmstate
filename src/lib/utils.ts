import { Address } from "tevm";

export const uniqueAddresses = (addresses: Array<Address | undefined>): Array<Address> => {
  let existingAddresses = new Set<string>();

  return addresses.filter((address) => {
    if (!address || existingAddresses.has(address.toLowerCase())) return false;
    existingAddresses.add(address.toLowerCase());
    return true;
  }) as Address[];
};
