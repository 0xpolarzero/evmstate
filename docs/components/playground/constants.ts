import { EthjsAddress } from "tevm/utils";
import type { ExtractAbiFunctions } from "abitype";

import * as LAYOUTS from "@test/codegen/layouts/index.js";
import * as CONTRACTS from "@test/contracts/index.js";

export const contract = CONTRACTS["Playground"].withAddress(
  EthjsAddress.fromString("0x987C2AF139EAEaBdF8D6d3d1723C1883bEa1f2AF").toString(),
);
export const layout = LAYOUTS["Playground"];
export const callerAddress = EthjsAddress.fromString("0xCa11e40000000000000000000000000000000000");
export const localStorageKey = "EVMSTATE_PLAYGROUND_STATE";

export const functionDescriptions: Record<ExtractAbiFunctions<typeof contract.abi, "nonpayable">["name"], string> = {
  addValue: "Adds to a dynamic array",
  addUser: "Adds a user struct to a mapping and array",
  toggleUserActive: "Toggles a boolean in a struct within a mapping",
  setBalance: "Updates a simple mapping value",
  setAllowance: "Updates a nested mapping value",
  addTransaction: "Adds to a dynamic array within a mapping",
  updateBasicValues: "Updates primitive types (uint256, bool)",
  updatePackedValues: "Updates packed storage variables (uint8, uint16, uint32, bool)",
  setStringAndBytes: "Updates string and bytes storage",
  setFixedValue: "Updates a fixed-size array element",
};
