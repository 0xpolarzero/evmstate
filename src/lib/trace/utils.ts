import {
  Abi,
  BlockTag,
  ContractFunctionName,
  createMemoryClient,
  encodeFunctionData,
  Hex,
  http,
  MemoryClient,
} from "tevm";
import { Common } from "tevm/common";

import { debug } from "@/debug";
import { ExploreStorageConfig } from "@/lib/explore/config";
import { TraceStorageAccessOptions, TraceStorageAccessTxParams, TraceStorageAccessTxWithData } from "@/lib/trace/types";

/** Creates a Tevm client from the provided options */
export const createClient = (options: {
  rpcUrl?: string;
  common?: Common;
  blockTag?: BlockTag | bigint;
}): MemoryClient => {
  const { rpcUrl, common, blockTag } = options;
  if (!rpcUrl) throw new Error("You need to provide a rpcUrl if you don't provide a client directly");

  return createMemoryClient({
    common,
    fork: {
      transport: http(rpcUrl),
      blockTag: blockTag ?? "latest",
    },
    miningConfig: { type: "manual" },
  });
};

export const getUnifiedParams = async <
  TAbi extends Abi | readonly unknown[] = Abi,
  TFunctionName extends ContractFunctionName<TAbi> = ContractFunctionName<TAbi>,
>(
  args: TraceStorageAccessOptions & TraceStorageAccessTxParams<TAbi, TFunctionName> & { config?: ExploreStorageConfig },
): Promise<TraceStorageAccessTxWithData & { client: MemoryClient }> => {
  const { client: _client, rpcUrl, common } = args;

  // Create the tevm client
  const client = _client ?? createClient({ rpcUrl, common });

  // Return early if the tx was already provided in calldata format
  if (args.from && args.data) return { client, from: args.from, to: args.to, data: args.data, value: args.value };

  // Encode calldata if the contract call was provided (abi, functionName, args)
  if (args.from && args.to && args.abi && args.functionName && args.args) {
    try {
      // @ts-expect-error complex union type not exactly similar
      const data = encodeFunctionData(args);
      return { client, from: args.from, to: args.to, data, value: args.value };
    } catch (err) {
      debug(`Failed to encode function data: ${err}`);
      throw err;
    }
  }

  // If we provide a value but no txHash, it's a simple transfer
  if (args.value && !args.txHash) return { client, from: args.from, to: args.to, value: args.value };

  // In this case, we need to replay the transaction
  if (!args.txHash)
    throw new Error("You need to provide a txHash if you don't provide the transaction data or contract call");

  // If we're replaying a transaction, extract the from, to, and data from the transaction
  try {
    const tx = await client.getTransaction({ hash: args.txHash });

    // Duplicate the client but fork before the transaction's block
    const clientBeforeTx = createMemoryClient({
      ...client,
      fork: {
        transport: client.transport,
        // Ideally we'd want to fork just right before that tx as some same-block txs can impact the state
        blockTag: tx.blockNumber > 0 ? tx.blockNumber - BigInt(1) : BigInt(0),
      },
    });

    return {
      client: clientBeforeTx,
      from: tx.from,
      to: tx.to ?? undefined,
      // TODO: remove when correctly formatted (tx in block mined here has data instead of input)
      // @ts-expect-error Property 'data' does not exist on type Transaction
      data: tx.input ? (tx.input as Hex) : (tx.data as Hex),
      value: tx.value,
    };
  } catch (err) {
    debug(`Failed to get transaction for replaying ${args.txHash}: ${err}`);
    throw err;
  }
};

/** A helper function to clean up trace objects by removing undefined or zero values */
export const cleanTrace = (obj: any) => {
  const { current, next, note, ...rest } = obj;
  let trace = { ...rest };

  // Only include note if it exists
  if (note) trace.note = note;

  // Same for current and next
  trace.current = { hex: current.hex };
  if (current.decoded !== undefined) trace.current = { hex: current.hex, decoded: current.decoded };

  if (next && rest.modified) {
    trace.next = { hex: next.hex };
    if (next.decoded !== undefined) trace.next = { hex: next.hex, decoded: next.decoded };
  }

  return trace;
};
