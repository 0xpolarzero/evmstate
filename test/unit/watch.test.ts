import { assert, describe, expect, it } from "vitest";

import { ACCOUNTS, CONTRACTS, LAYOUTS } from "@test/constants.js";
import { getClient, waitFor } from "@test/utils.js";
import { watchState } from "@/lib/watch/index.js";
import type { StateChange } from "@/lib/watch/types.js";

const { StoragePacking } = CONTRACTS;
const { caller, recipient } = ACCOUNTS;

/**
 * Contract watcher tests
 *
 * This test suite verifies:
 *
 * - Watching a contract for storage changes
 * - Watching an EOA for state changes
 */

describe("Watching a contract", () => {
  it("should be able to watch a contract for state changes", async () => {
    const client = getClient();
    const stateChanges: Array<StateChange<typeof LAYOUTS.StoragePacking>> = [];

    const unsubscribe = await watchState({
      client,
      address: StoragePacking.address,
      abi: StoragePacking.abi,
      storageLayout: LAYOUTS.StoragePacking,
      onStateChange: (state) => stateChanges.push(state),
    });

    // Run a transaction that modifies the contract state
    const { txHash } = await client.tevmContract({
      ...StoragePacking.write.setSmallValues(42, 123, true, caller.toString()),
      from: caller.toString(),
      addToBlockchain: true,
    });
    assert(txHash, "txHash is required");

    await waitFor(() => stateChanges.length > 0, { timeout: 5_000 });
    expect(stateChanges).toMatchSnapshot();

    unsubscribe();
  });

  it("should be able to watch an EOA for state changes", async () => {
    const client = getClient();
    const stateChanges: Array<StateChange> = [];

    const unsubscribe = await watchState({
      client,
      address: recipient.toString(),
      onStateChange: (state) => stateChanges.push(state),
    });

    // Run a transaction that modifies the contract state
    const { txHash } = await client.tevmCall({
      from: caller.toString(),
      to: recipient.toString(),
      value: 100n,
      addToBlockchain: true,
    });
    assert(txHash, "txHash is required");

    await waitFor(() => stateChanges.length > 0, { timeout: 5_000 });
    expect(stateChanges).toMatchSnapshot();

    unsubscribe();
  });
});
