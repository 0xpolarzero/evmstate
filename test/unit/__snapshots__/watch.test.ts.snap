// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Watching a contract > should be able to watch a contract for state changes 1`] = `
[
  {
    "balance": {
      "current": 0n,
      "modified": false,
    },
    "code": {
      "current": "0x608060405234801561000f575f5ffd5b50600436106100a7575f3560e01c80634baa58331161006f5780634baa58331461013e57806361d4b2671461015a578063822d530f1461017a578063e5008f5a14610196578063f0827988146101b2578063fcfe64a3146101ce576100a7565b8063138b2a8a146100ab5780632c60e58a146100c757806334a062fe146100e35780633bc5de301461010457806341657ee614610122575b5f5ffd5b6100c560048036038101906100c091906104ab565b6101ec565b005b6100e160048036038101906100dc919061059b565b61020b565b005b6100eb61029d565b6040516100fb949392919061062c565b60405180910390f35b61010c6102fd565b6040516101199190610687565b60405180910390f35b61013c600480360381019061013791906106ca565b610306565b005b61015860048036038101906101539190610761565b610310565b005b61016261038e565b60405161017193929190610818565b60405180910390f35b610194600480360381019061018f919061084d565b6103d6565b005b6101b060048036038101906101ab9190610878565b6103e0565b005b6101cc60048036038101906101c791906108cd565b6103fc565b005b6101d6610467565b6040516101e3919061092c565b60405180910390f35b8060035f6101000a81548161ffff021916908361ffff16021790555050565b835f5f6101000a81548160ff021916908360ff160217905550825f60016101000a81548160ff021916908360ff160217905550815f60026101000a81548160ff021916908315150217905550805f60036101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff16021790555050505050565b5f5f5f5f5f5f9054906101000a900460ff165f60019054906101000a900460ff165f60029054906101000a900460ff165f60039054906101000a900473ffffffffffffffffffffffffffffffffffffffff16935093509350935090919293565b5f600254905090565b8060028190555050565b845f5f6101000a81548160ff021916908360ff160217905550835f60016101000a81548160ff021916908360ff1602179055508260035f6101000a81548161ffff021916908361ffff16021790555081600360026101000a81548163ffffffff021916908363ffffffff160217905550806001819055505050505050565b5f5f5f60035f9054906101000a900461ffff16600360029054906101000a900463ffffffff16600360069054906101000a900467ffffffffffffffff16925092509250909192565b8060018190555050565b805f5f6101000a81548160ff021916908360ff16021790555050565b8260035f6101000a81548161ffff021916908361ffff16021790555081600360026101000a81548163ffffffff021916908363ffffffff16021790555080600360066101000a81548167ffffffffffffffff021916908367ffffffffffffffff160217905550505050565b5f600154905090565b5f5ffd5b5f61ffff82169050919050565b61048a81610474565b8114610494575f5ffd5b50565b5f813590506104a581610481565b92915050565b5f602082840312156104c0576104bf610470565b5b5f6104cd84828501610497565b91505092915050565b5f60ff82169050919050565b6104eb816104d6565b81146104f5575f5ffd5b50565b5f81359050610506816104e2565b92915050565b5f8115159050919050565b6105208161050c565b811461052a575f5ffd5b50565b5f8135905061053b81610517565b92915050565b5f73ffffffffffffffffffffffffffffffffffffffff82169050919050565b5f61056a82610541565b9050919050565b61057a81610560565b8114610584575f5ffd5b50565b5f8135905061059581610571565b92915050565b5f5f5f5f608085870312156105b3576105b2610470565b5b5f6105c0878288016104f8565b94505060206105d1878288016104f8565b93505060406105e28782880161052d565b92505060606105f387828801610587565b91505092959194509250565b610608816104d6565b82525050565b6106178161050c565b82525050565b61062681610560565b82525050565b5f60808201905061063f5f8301876105ff565b61064c60208301866105ff565b610659604083018561060e565b610666606083018461061d565b95945050505050565b5f819050919050565b6106818161066f565b82525050565b5f60208201905061069a5f830184610678565b92915050565b6106a98161066f565b81146106b3575f5ffd5b50565b5f813590506106c4816106a0565b92915050565b5f602082840312156106df576106de610470565b5b5f6106ec848285016106b6565b91505092915050565b5f63ffffffff82169050919050565b61070d816106f5565b8114610717575f5ffd5b50565b5f8135905061072881610704565b92915050565b5f819050919050565b6107408161072e565b811461074a575f5ffd5b50565b5f8135905061075b81610737565b92915050565b5f5f5f5f5f60a0868803121561077a57610779610470565b5b5f610787888289016104f8565b9550506020610798888289016104f8565b94505060406107a988828901610497565b93505060606107ba8882890161071a565b92505060806107cb8882890161074d565b9150509295509295909350565b6107e181610474565b82525050565b6107f0816106f5565b82525050565b5f67ffffffffffffffff82169050919050565b610812816107f6565b82525050565b5f60608201905061082b5f8301866107d8565b61083860208301856107e7565b6108456040830184610809565b949350505050565b5f6020828403121561086257610861610470565b5b5f61086f8482850161074d565b91505092915050565b5f6020828403121561088d5761088c610470565b5b5f61089a848285016104f8565b91505092915050565b6108ac816107f6565b81146108b6575f5ffd5b50565b5f813590506108c7816108a3565b92915050565b5f5f5f606084860312156108e4576108e3610470565b5b5f6108f186828701610497565b93505060206109028682870161071a565b9250506040610913868287016108b9565b9150509250925092565b6109268161072e565b82525050565b5f60208201905061093f5f83018461091d565b9291505056fe0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": false,
    },
    "storage": {
      "flag": {
        "kind": "primitive",
        "name": "flag",
        "trace": [
          {
            "current": {
              "decoded": false,
              "hex": "0x00",
            },
            "fullExpression": "flag",
            "modified": true,
            "next": {
              "decoded": true,
              "hex": "0x01",
            },
            "path": [],
            "slots": [
              "0x0000000000000000000000000000000000000000000000000000000000000000",
            ],
          },
        ],
        "type": "bool",
      },
      "smallValue1": {
        "kind": "primitive",
        "name": "smallValue1",
        "trace": [
          {
            "current": {
              "decoded": 0,
              "hex": "0x00",
            },
            "fullExpression": "smallValue1",
            "modified": true,
            "next": {
              "decoded": 42,
              "hex": "0x2a",
            },
            "path": [],
            "slots": [
              "0x0000000000000000000000000000000000000000000000000000000000000000",
            ],
          },
        ],
        "type": "uint8",
      },
      "smallValue2": {
        "kind": "primitive",
        "name": "smallValue2",
        "trace": [
          {
            "current": {
              "decoded": 0,
              "hex": "0x00",
            },
            "fullExpression": "smallValue2",
            "modified": true,
            "next": {
              "decoded": 123,
              "hex": "0x7b",
            },
            "path": [],
            "slots": [
              "0x0000000000000000000000000000000000000000000000000000000000000000",
            ],
          },
        ],
        "type": "uint8",
      },
      "someAddress": {
        "kind": "primitive",
        "name": "someAddress",
        "trace": [
          {
            "current": {
              "decoded": "0x0000000000000000000000000000000000000000",
              "hex": "0x00",
            },
            "fullExpression": "someAddress",
            "modified": true,
            "next": {
              "decoded": "0x0000000000000000000000000000000000000001",
              "hex": "0x0000000000000000000000000000000000000001",
            },
            "path": [],
            "slots": [
              "0x0000000000000000000000000000000000000000000000000000000000000000",
            ],
          },
        ],
        "type": "address",
      },
    },
    "txHash": "0x9cbddc51a5e5b001d817bfa10aaf35e948caf395173c86c32d63217284c72efb",
  },
]
`;

exports[`Watching a contract > should be able to watch an EOA for state changes 1`] = `
[
  {
    "balance": {
      "current": 10000000000000000000n,
      "modified": true,
      "next": 10000000000000000100n,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": false,
    },
    "storage": {},
    "txHash": "0xd6c89428e3dda7c1a4c419050436ad8e7d3641d62a066a9386b8d7eb59d64993",
  },
]
`;
